### PR TITLE
Add naming conventions for collections - plurality

### PR DESCRIPTION
**Pull request by jerzyn**
Add description for naming conventions for collections - REST APIs

### CODE REVIEW

File: rest-api-guidelines/quality/evolution/naming-conventions.md [hunk 1]
```diff
@@ -13,6 +13,14 @@ Every identifier **MUST** be in American English and written in `lowercase`. An
 
 Every URI **MUST** follow the General Rules except for the `camelCase` rule. Instead, a hyphen \(`-`\) **SHOULD** be used to delimit combined words \(kebab-case\). Besides, a URI **MUST NOT** end with a trailing slash \(`/`\).
 
+Plural nouns SHOULD be used in the URI where appropriate to identify collections of data resources (e.g. ``/orders``, ``/products``).
```
Comment (zdne) [inline rest-api-guidelines/quality/evolution/naming-conventions.md]: minor: formatting – it seems that you have double backticks there - ` `` ` is it intentional or should it be just one?

File: rest-api-guidelines/quality/evolution/naming-conventions.md [hunk 2]
```diff
@@ -13,6 +13,14 @@ Every identifier **MUST** be in American English and written in `lowercase`. An
...
+
+An individual resource in a collection of resources MAY exist directly beneath the collection URI.  (e.g. ``/orders/{order_id}``).
+
+Sub-resource collections MAY exist directly beneath an individual resource. This should convey a relationship to another collection of resources (e.g. ``/orders/{order_id}/items``).
```
Comment (zdne) [inline rest-api-guidelines/quality/evolution/naming-conventions.md]: In general, I a not a fan of policing URL structures and inferring any semantics from the URL structure - e.g. "A is s a sub-resource of B because I am guessing it from the URL of `/A/B`". 

Per RFC 7320: 

https://tools.ietf.org/html/rfc7320

> The most successful strategy for applications and extensions that wish to use URIs is to use them in the fashion they were designed: as links that are exchanged as part of the protocol, rather than statically specified syntax.

With that being said, I see a good intention in this PR and as such, I am hesitant to refuse it, but also to merge it. Let's have some discusson about it :)

File: rest-api-guidelines/quality/evolution/naming-conventions.md [hunk 3]
```diff
@@ -13,6 +13,14 @@ Every identifier **MUST** be in American English and written in `lowercase`. An
...
+
+Sub-resource individual resources MAY exist, but should be avoided in favor of top-level resources. (e.g. ``/orders/{order_id}/items/{item_id}``).
```
Comment (zdne) [inline rest-api-guidelines/quality/evolution/naming-conventions.md]: Per my comment above, I do not think this is a good recommendation even as a rule of thumb. If you want to offer an that "lists an item by its id on a particular order" This is the best way to express it, even if the items ids are unique across orders...

File: rest-api-guidelines/quality/evolution/naming-conventions.md (hunk 1 above)
Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: it should be just `code`, might be an typo

File: rest-api-guidelines/quality/evolution/naming-conventions.md (hunk 3 above)
Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: im not sure if i got this. can you rephrase please?

File: rest-api-guidelines/quality/evolution/naming-conventions.md (hunk 2 above)
Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: the intention of this PR comes from the fact, that I see a lot of APIs which have a semantically incoherent structure and/or have endpoints like /member/bulk, etc. And yes, I agree, that it may suggest a parent/child structure in some cases. On the other hand how would you express a relations between some resources?

File: rest-api-guidelines/quality/evolution/naming-conventions.md (hunk 3 above)
Comment (zdne) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: @jerzyn lets discuss it over a short call :) It might be more effective than here.

Comment (zwedoc) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: Hi @zdne, @jerzyn I think I can agree with you both :)

I agree with @jerzyn as I think it is always good to explicitly express hard relations between root resources and sub-resources as orders and lineItem. Following with this example, if a lineItem does not have a meaning/lifecycle outside the context of an order, the recommendation should be to not expose that sub-resource with a root path and always use the root resource to navigate to it (via links or embedded attribute). That's btw inline with the Aggregate's approach in DDD and transactional boundaries. For instance, Can I update a lineItem without affecting the total amount of an order? If there is a hard relation URLs SHOULD make it explicit. 

However I do also agree with @zdne on making it a rule of thumb or not as it be confusing when creating compositions that aggregate resource in a way that is valid for a specific use case but don't necessarily reflect with exactness the relation between resources.

Comment (zdne) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: > Following with this example, if a lineItem does not have a meaning/lifecycle outside the context of an order, the recommendation should be to not expose that sub-resource with a root path and always use the root resource to navigate to it. 

I like this as a rule of a thumb.

Comment (zdne) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: > If there is a hard relation URLs SHOULD make it explicit.

Something is fighting in me about this – the URLs should really be the identifiers in the first place – one should not guess the relations types from URLs, yet many ppl are still doing it. Frankly, If I would see a Hypermedia API that has URLs in the form of UUIDs (e.g. `/00AA1FF11D2310` and `/112AAAA00003`) I would be fine with it ...

File: rest-api-guidelines/quality/evolution/naming-conventions.md (hunk 2 above)
Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: Are we good if I delete the sub-resources part and leave only the part on individual resource in a collection @zdne ?

Comment (zdne) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: Yes please.

Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: done. Can I merge it?

File: rest-api-guidelines/quality/evolution/naming-conventions.md [hunk 4]
```diff
@@ -13,6 +13,10 @@ Every identifier **MUST** be in American English and written in `lowercase`. An
 
 Every URI **MUST** follow the General Rules except for the `camelCase` rule. Instead, a hyphen \(`-`\) **SHOULD** be used to delimit combined words \(kebab-case\). Besides, a URI **MUST NOT** end with a trailing slash \(`/`\).
 
+Plural nouns SHOULD be used in the URI where appropriate to identify collections of data resources (e.g. ``/orders``, ``/products``).
```
Comment (zdne) [inline rest-api-guidelines/quality/evolution/naming-conventions.md]: are the double backquotes intentional?

Comment (jerzyn) [reply rest-api-guidelines/quality/evolution/naming-conventions.md]: no, a typo!

File: rest-api-guidelines/quality/evolution/naming-conventions.md [hunk 5]
```diff
@@ -13,6 +13,10 @@ Every identifier **MUST** be in American English and written in `lowercase`. An
...
+
+An individual resource in a collection of resources MAY exist directly beneath the collection URI.  (e.g. ``/orders/{order_id}``).
```
Comment (zdne) [inline rest-api-guidelines/quality/evolution/naming-conventions.md]: Again I feel this is not bringing any value – of course they may exist there, as a matter of fact where else they would be?

**Review by zdne — DISMISSED** [review]

**Review by Amzani — APPROVED** [review]

**Review by Amzani — APPROVED** [review]

**Review by jesusjavierdediego — APPROVED** [review]
